File: python/generate_test_data.c
/*
 * 生成测试数据并保存到C语言头文件中
 * 与Python脚本生成相同的数据用于对比
 */

#include <stdio.h>
#include <math.h>

#define SAMPLE_RATE 16000
#define DURATION 0.2
#define FREQUENCY 500
#define AMPLITUDE 0.3
#define SAMPLES ((int)(SAMPLE_RATE * DURATION))

static float test_input_signal[SAMPLES];

void generate_sine_wave() {
    // 生成500Hz正弦波测试信号
    for (int i = 0; i < SAMPLES; i++) {
        float t = i / (float)SAMPLE_RATE;
        test_input_signal[i] = AMPLITUDE * sinf(2.0f * M_PI * FREQUENCY * t);
    }
}

void save_to_header_file() {
    FILE* f = fopen("test_data_c.h", "w");
    if (!f) {
        printf("无法创建文件 test_data_c.h\n");
        return;
    }
    
    fprintf(f, "/*\n");
    fprintf(f, " * 测试数据头文件 (C版本)\n");
    fprintf(f, " * 自动生成的正弦波测试信号\n");
    fprintf(f, " */\n\n");
    fprintf(f, "#ifndef TEST_DATA_C_H\n");
    fprintf(f, "#define TEST_DATA_C_H\n\n");
    fprintf(f, "#define TEST_SIGNAL_LENGTH %d\n\n", SAMPLES);
    fprintf(f, "static const float test_input_signal_c[TEST_SIGNAL_LENGTH] = {\n");
    
    // 每行写入8个值
    for (int i = 0; i < SAMPLES; i += 8) {
        int remaining = SAMPLES - i;
        int count = remaining < 8 ? remaining : 8;
        
        fprintf(f, "    ");
        for (int j = 0; j < count; j++) {
            fprintf(f, "%.6ff", test_input_signal[i + j]);
            if (j < count - 1) {
                fprintf(f, ", ");
            }
        }
        
        if (i + 8 < SAMPLES) {
            fprintf(f, ",\n");
        } else {
            fprintf(f, "\n");
        }
    }
    
    fprintf(f, "};\n\n");
    fprintf(f, "#endif // TEST_DATA_C_H\n");
    
    fclose(f);
    printf("测试数据已保存到 test_data_c.h\n");
}

void print_signal_info() {
    printf("生成测试数据:\n");
    printf("采样率: %d Hz\n", SAMPLE_RATE);
    printf("持续时间: %.1f 秒\n", DURATION);
    printf("频率: %d Hz\n", FREQUENCY);
    printf("幅度: %.1f\n", AMPLITUDE);
    printf("采样点数: %d\n", SAMPLES);
    printf("\n");
    
    // 显示前10个采样点
    printf("前10个采样点:\n");
    for (int i = 0; i < 10 && i < SAMPLES; i++) {
        printf("  [%d]: %.6f\n", i, test_input_signal[i]);
    }
    printf("\n");
}

int main() {
    generate_sine_wave();
    print_signal_info();
    save_to_header_file();
    printf("完成!\n");
    return 0;
}

File: python/generate_test_data.py
"""
生成测试数据并保存到C语言头文件中
"""

import numpy as np

def generate_sine_wave():
    """生成500Hz正弦波测试信号"""
    # 参数设置
    sr = 16000        # 采样率 16kHz
    duration = 0.2    # 持续时间 200ms
    frequency = 500   # 频率 500Hz
    amplitude = 0.3   # 幅度 0.3
    
    # 生成测试信号 (3200个采样点)
    t = np.linspace(0, duration, int(sr * duration), endpoint=False)
    test_signal = amplitude * np.sin(2 * np.pi * frequency * t)
    
    return test_signal

def save_to_header_file(signal, filename):
    """将信号保存到C语言头文件"""
    with open(filename, "w") as f:
        f.write("/*\n")
        f.write(" * 测试数据头文件\n")
        f.write(" * 自动生成的正弦波测试信号\n")
        f.write(" */\n\n")
        f.write("#ifndef TEST_DATA_H\n")
        f.write("#define TEST_DATA_H\n\n")
        f.write(f"#define TEST_SIGNAL_LENGTH {len(signal)}\n\n")
        f.write("static const float test_input_signal[TEST_SIGNAL_LENGTH] = {\n")
        
        # 每行写入8个值
        for i in range(0, len(signal), 8):
            values = signal[i:i+8]
            line = ", ".join([f"{val:.6f}f" for val in values])
            f.write(f"    {line}")
            if i + 8 < len(signal):
                f.write(",\n")
            else:
                f.write("\n")
        
        f.write("};\n\n")
        f.write("#endif // TEST_DATA_H\n")
    
    print(f"测试数据已保存到 {filename}")

def main():
    # 生成正弦波测试信号
    test_signal = generate_sine_wave()
    
    # 显示一些统计信息
    print("生成测试数据:")
    print(f"采样率: 16000 Hz")
    print(f"持续时间: 0.2 秒")
    print(f"频率: 500 Hz")
    print(f"幅度: 0.3")
    print(f"采样点数: {len(test_signal)}")
    print()
    
    # 显示前10个采样点
    print("前10个采样点:")
    for i in range(10):
        print(f"  [{i}]: {test_signal[i]:.6f}")
    print()
    
    # 保存到头文件
    save_to_header_file(test_signal, "test_data.h")
    
    print("完成!")

if __name__ == "__main__":
    main()